control-plane/agents/src/bin/core/controller/reconciler/node/snapshot.rs

```diff
@@ -0,0 +1,105 @@
+use tracing::info;
+
+/// Node snapshot garbage collector.
+#[derive(Debug)]
+pub(super) struct NodeSnapshotGarbageCollector {
+    counter: PollTimer,
+}
+impl NodeSnapshotGarbageCollector {
+    /// Return a new `Self`.
+    pub(super) fn new() -> Self {
+        Self {
+            counter: ReconcilerWorker::garbage_collection_period(),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl TaskPoller for NodeSnapshotGarbageCollector {
+    async fn poll(&mut self, context: &PollContext) -> PollResult {
+        let nodes = context.specs().nodes();
+        let mut results = Vec::with_capacity(nodes.len());
+
+        for node in nodes {
+            let mut node_spec = context.registry().specs().guarded_node(node.id()).await?;
+            results.push(delete_unknown_volume_snapshot_reconciler(context, &mut node_spec).await);
+        }
+        Self::squash_results(results)
+    }
+
+    async fn poll_timer(&mut self, _context: &PollContext) -> bool {
+        self.counter.poll()
+    }
+}
+
+/// Identify unknown snapshots and attempt to clean them up.
+async fn delete_unknown_volume_snapshot_reconciler(
+    context: &PollContext,
+    node_spec: &mut OperationGuardArc<NodeSpec>,
+) -> PollResult {
+    let node_wrapper = match context.registry().node_wrapper(node_spec.uid()).await {
+        Ok(node_wrapper) => node_wrapper,
+        Err(error) => return PollResult::Err(error),
+    };
+
+    if let Ok(snaps) = node_wrapper
+        .list_repl_snapshots(&ListReplicaSnapshots::All)
+        .await
+    {
+        for snap in snaps {
+            if let Ok((snap_uuid, txn_id)) = extract_uuid_txn_id(snap.snap_name()) {
+                // Don't cleanup if the snapshot is known, i.e part of present snapshot's
+                // transaction.
+                if let Some(snap_rsc) = context.registry().specs().volume_snapshot_rsc(&snap_uuid) {
+                    let snapshot_spec = snap_rsc.lock().clone();
+                    if snapshot_spec.metadata().transactions().contains_key(txn_id) {
+                        continue;
+                    }
+                }
+
+                // Attempt destroy of the snapshot if it qualifies as unknown.
+                if node_wrapper
+                    .destroy_repl_snapshot(&DestroyReplicaSnapshot::new(
+                        snap.snap_uuid().clone(),
+                        snap.pool_uuid().clone(),
+                    ))
+                    .await
+                    .is_ok()
+                {
+                    info!(
+                        "Deleted unidentified replica snapshot: {} from pool: {}",
+                        snap.snap_uuid(),
+                        snap.snap_uuid()
+                    );
+                }
+            }
+        }
+    }
+
+    Ok(PollerState::Idle)
+}
+
+/// Extracts the uuid and txn id from the snapshot name of format <snap_uuid>/<txn_id>.
+fn extract_uuid_txn_id(input: &str) -> Result<(SnapshotId, &str), SvcError> {
```

Comment:
Just make it an Option here since it's not an error if snapshots were created outside of control-plane.